galley/pkg/config/processing/runtime.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processing
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"istio.io/api/mesh/v1alpha1"
+
+	"istio.io/istio/galley/pkg/config/event"
+)
+
+// ProcessorOptions are options that are passed to event.Processors during startup.
+type ProcessorOptions struct {
+	MeshConfig   *v1alpha1.MeshConfig
+	DomainSuffix string
+}
+
+// RuntimeOptions is options for Runtime
+type RuntimeOptions struct {
+	Sources      []event.Source
+	Processor    event.Processor
+	DomainSuffix string
+}
+
+// Clone returns a cloned copy of the RuntimeOptions.
+func (o *RuntimeOptions) Clone() RuntimeOptions {
+	sources := make([]event.Source, len(o.Sources))
+	copy(sources, o.Sources)
+
+	return RuntimeOptions{
+		Sources:      sources,
+		Processor:    o.Processor,
+		DomainSuffix: o.DomainSuffix,
+	}
+}
+
+// Runtime is the config processing runtime.
+type Runtime struct { // nolint:maligned
+	mu sync.RWMutex
+
+	// counter for session id. The current value reflects the processing session's id.
+	sessionIDCtr int32
+
+	// runtime options that was passed as parameters to the command-line.
+	options RuntimeOptions
+
+	// stopCh is used send stop signal completion to the background go-routine.
+	stopCh chan struct{}
+
+	// wg is used to synchronize the completion of Stop call with the completion of the background
+	// go routine.
+	wg      sync.WaitGroup
+	session atomic.Value
+}
+
+// NewRuntime returns a new instance of a processing.Runtime.
+func NewRuntime(o RuntimeOptions) *Runtime {
+
+	r := &Runtime{
+		options: o.Clone(),
+	}
+
+	h := event.HandlerFromFn(r.handle)
+	for _, s := range o.Sources {
+		s.Dispatch(h)
```

Comment:
nit: `Dispatch` sounds like it would pass an event to `h` now. This is really more of a registration ... maybe something like `DispatchTo` or `Register` would be more clear?

Reply:
Which golang interface are you referring to exactly? IIUC what you're doing here would be more akin to `Watch`, no?

Reply:
https://github.com/istio/istio/blob/dbbe3009d145edd0b3e3bef1044e6331fdea6aaf/galley/pkg/config/event/dispatcher.go#L18

Which is implemented by Source.

Yes, this is similar to Watch. Just to be clear, I am happy to change the name here to something better, if you have a better suggestion. I am not happy with the current name.